show image vulnerability in status under project details

nemesis09 opened the pull request:
**Fixes:**
https://issues.redhat.com/browse/ODC-4881

**Description:**
- Show Image Vulnerability in status under Project Details for each Project

**Screens:**
![Screenshot from 2020-10-15 10-06-21](https://user-images.githubusercontent.com/38663217/96077758-434b6b00-0ece-11eb-9e64-2ad767d67097.png)
![Screenshot from 2020-10-15 10-06-30](https://user-images.githubusercontent.com/38663217/96077762-45adc500-0ece-11eb-969e-561538f0d59e.png)
![Screenshot from 2020-10-15 10-06-54](https://user-images.githubusercontent.com/38663217/96077763-47778880-0ece-11eb-9f93-b592e9e08cb2.png)
![Screenshot from 2020-10-15 10-07-06](https://user-images.githubusercontent.com/38663217/96077770-48a8b580-0ece-11eb-89cf-839da01969a7.png)
![Screenshot from 2020-11-03 20-03-26](https://user-images.githubusercontent.com/38663217/97998126-bbdb8280-1e0f-11eb-8566-3b5f7ef10195.png)
![image-vulnerability](https://user-images.githubusercontent.com/38663217/98902043-ee097600-24da-11eb-94db-350e8d95ada4.gif)

**Test Coverage:**
![Screenshot from 2020-10-15 10-16-36](https://user-images.githubusercontent.com/38663217/96078299-8b1ec200-0ecf-11eb-9713-46e202e67945.png)

**Browser Conformance:**
- [x] Firefox
- [ ] Chrome
- [ ] Safari
- [ ] Edge

invincibleJai commented:
@nemesis09 the position of popover arrow doesn't seem to be right

![image](https://user-images.githubusercontent.com/5129024/96079050-73e0d400-0ed1-11eb-9355-8342942c67db.png)

nemesis09 commented:
> @nemesis09 the position of popover arrow doesn't seem to be right
> 
> ![image](https://user-images.githubusercontent.com/5129024/96079050-73e0d400-0ed1-11eb-9355-8342942c67db.png)

this seems to be an existing bug. looking into it. thanks!

invincibleJai commented on frontend/packages/container-security/src/components/summary.tsx:
```diff
@@ -130,14 +140,23 @@ export const SecurityBreakdownPopup: React.FC<SecurityBreakdownPopupProps> = ({
-                      )}
+                      {namespace
+                        ? imageNameClamped(
+                            resource.find(({ metadata }) => metadata.name === v.metadata.name).spec
```
nit: we need to consider the case if find returns undefined

nemesis09 replied:
i figured there shouldn't be a case where `find` returns undefined since `v` is an element from `fixableVulns` which is a subset of `resource`, and there is a check for `fixableVulns` being empty which prevents execution when `fixableVulns` is empty

nemesis09 replied:
i figured there shouldn't be a case where `find` returns undefined since `v` is an element from `fixableVulns` which is a subset of `resource`, and there is a check for `fixableVulns` being empty which prevents execution when `fixableVulns` is empty

invincibleJai commented on frontend/packages/container-security/src/components/summary.tsx:
```diff
@@ -130,14 +140,23 @@ export const SecurityBreakdownPopup: React.FC<SecurityBreakdownPopupProps> = ({
-                    <ExternalLink href={quayURLFor(v)} text={`${v.status?.fixableCount} fixable`} />
+                    <ExternalLink
+                      href={quayURLFor(v)}
+                      text={`${getVulnerabilityCountText(v, namespace)} fixable`}
```
as you added util here m shouldn't we return `fixable` as well from util now

nemesis09 replied:
done

invincibleJai commented on frontend/public/components/dashboard/dashboards-page/cluster-dashboard/health-item.tsx:
```diff
@@ -221,10 +222,18 @@ export const PrometheusHealthItem = withDashboardResources<PrometheusHealthItemP
-  const resourcesResult = useK8sWatchResources(resources);
-
+  const resourcesResultUnfiltered = useK8sWatchResources(resources);
+  const resourceResultFilteredByNamespace = _.set(
+    _.cloneDeep(resourcesResultUnfiltered),
+    'imageManifestVuln.data',
+    _.filter(
+      resourcesResultUnfiltered.imageManifestVuln.data,
+      (imgvuln: any) => imgvuln?.metadata?.namespace === namespace,
+    ),
+  );
```
do we need this?

can we update the resources itself with namespace if have one so that we'll get appropriate data. WDYT?

invincibleJai commented on frontend/public/components/dashboard/project-dashboard/status-card.scss:
```diff
@@ -0,0 +1,9 @@
+.co-project-dashboard__status {
+  display: flex;
+  align-items: center;
+  padding-right: var(--pf-global--spacer--xl);;
```
extra `;`

nemesis09 replied:
fixed

invincibleJai commented on frontend/public/components/dashboard/project-dashboard/status-card.tsx:
```diff
@@ -1,22 +1,64 @@
+  isDashboardsOverviewHealthSubsystem,
+  useExtensions,
+} from '@console/plugin-sdk';
+import { ContainerSecurityFlag } from '@console/container-security/src/const';
```
we should not import from `container-security` in public

nemesis09 replied:
fixed

invincibleJai commented on frontend/public/components/dashboard/project-dashboard/status-card.tsx:
```diff
@@ -1,22 +1,64 @@
+import { ContainerSecurityFlag } from '@console/container-security/src/const';
 import { ProjectDashboardContext } from './project-dashboard-context';
+import { ResourceHealthItem } from '../dashboards-page/cluster-dashboard/health-item';
+import './status-card.scss';
```
good to have line break before style imports

nemesis09 replied:
done

debsmita1 commented on frontend/public/components/dashboard/project-dashboard/status-card.tsx:
```diff
@@ -1,22 +1,64 @@
 export const StatusCard: React.FC = () => {
   const { obj } = React.useContext(ProjectDashboardContext);
+  const filterSubsystems = (subsystems: DashboardsOverviewHealthSubsystem[]) =>
+    subsystems.filter((s) => {
+      if (isDashboardsOverviewHealthResourceSubsystem(s)) {
+        return true;
+      }
+    });
```
```suggestion
    subsystems.filter(isDashboardsOverviewHealthResourceSubsystem);
```

nemesis09 replied:
done

vikram-raj reviewed (changes requested):
Clicking on a donut or on impact under popover is redirecting to all-namespace Vulnerabilities list page. It should redirect to the respective project Vulnerabilities list page.

nemesis09 commented:
> Clicking on a donut or on impact under popover is redirecting to all-namespace Vulnerabilities list page. It should redirect to the respective project Vulnerabilities list page.

updated the links to redirect to the vulnerabilities list page of the corresponding project.

invincibleJai commented on frontend/packages/container-security/src/components/summary.tsx:
```diff
@@ -44,14 +44,25 @@ export const quayURLFor = (vuln: ImageManifestVuln) => {
     .filter((v) => v.status?.fixableCount > 0)
     .reduce((all, v) => all.set(v.metadata.name, v), new Map<string, ImageManifestVuln>());
+  const getVulnerabilityCountText = (v: ImageManifestVuln, n?: string): string => {
+    const { highCount, mediumCount, lowCount, unknownCount } = v.status;
```
can `v.status` be undefined?
```suggestion
    const { highCount =0, mediumCount=0, lowCount=0, unknownCount=0} = v.status || {};
```

nemesis09 replied:
`v.status` shouldn't be undefined as it is a required property in the type `ImageManifestVuln`

invincibleJai commented on frontend/packages/container-security/src/components/summary.tsx:
```diff
@@ -44,14 +44,25 @@ export const quayURLFor = (vuln: ImageManifestVuln) => {
+  const getVulnerabilityCountText = (v: ImageManifestVuln, n?: string): string => {
+    const { highCount, mediumCount, lowCount, unknownCount } = v.status;
+    const totalCount =
+      (highCount ?? 0) + (mediumCount ?? 0) + (lowCount ?? 0) + (unknownCount ?? 0);
```
based on above

```suggestion
      highCount + mediumCount  + lowCount  + unknownCount;
```

nemesis09 replied:
done

invincibleJai commented on frontend/packages/container-security/src/components/summary.tsx:
```diff
@@ -44,14 +44,25 @@ export const quayURLFor = (vuln: ImageManifestVuln) => {
+    const { highCount, mediumCount, lowCount, unknownCount } = v.status;
+    const totalCount =
+      (highCount ?? 0) + (mediumCount ?? 0) + (lowCount ?? 0) + (unknownCount ?? 0);
+    return `${n ? `${v.status?.fixableCount} of ${totalCount}` : v.status?.fixableCount} fixable`;
```
nit: can destructure `fixableCount` above right?

nemesis09 replied:
destructured `fixableCount` with other properties

invincibleJai commented on frontend/public/components/dashboard/dashboards-page/cluster-dashboard/health-item.tsx:
```diff
@@ -226,11 +227,21 @@ export const PrometheusHealthItem = withDashboardResources<PrometheusHealthItemP
+          'imageManifestVuln.data',
+          _.filter(
+            resourcesResult.imageManifestVuln.data,
+            (imgvuln: any) => imgvuln?.metadata?.namespace === namespace,
```
do we know type here?

invincibleJai commented on frontend/public/components/dashboard/dashboards-page/cluster-dashboard/health-item.tsx:
```diff
@@ -226,11 +227,21 @@ export const PrometheusHealthItem = withDashboardResources<PrometheusHealthItemP
   const resourcesResult = useK8sWatchResources(resources);
-
-  const healthState = healthHandler(resourcesResult);
+  const filteredResourcesResult =
```
can't we get data/watch here `useK8sWatchResources` based on namespace as we already know that, we can avoid this. WDYT?

nemesis09 replied:
made resources to include namespace. This avoids the filter. Thanks for the catch.

invincibleJai reviewed (commented):
And one general comment , we should start support for i18n wrt changes done as part of this PR.

nemesis09 commented:
> And one general comment , we should start support for i18n wrt changes done as part of this PR.

Added support for i18n

vikram-raj commented on frontend/packages/container-security/src/components/summary.tsx:
```diff
@@ -77,19 +107,19 @@ export const SecurityBreakdownPopup: React.FC<SecurityBreakdownPopupProps> = ({
                           &nbsp;
-                          {_.uniqBy(vulnsFor(priority.value), 'metadata.name').length}{' '}
-                          {priority.title}
+                          {t('container-security~{{vulnImageCount}} {{priorityTitle}}', {
+                            vulnImageCount: _.uniqBy(vulnsFor(priority.value), 'metadata.name')
+                              .length,
+                            priorityTitle: priority.title,
+                          })}
```
Translation is not needed here. It won't have any effect.

nemesis09 replied:
this adds translation to the **Vulnerability Container Images** list. The list has items with the count of images with priorities like `1 High`. Do we not need translation for this? I have attached images for reference.
**Without translation**
![Screenshot from 2020-11-02 18-40-28](https://user-images.githubusercontent.com/38663217/97873042-0f30d080-1d3d-11eb-9a15-57d6f3eb404a.png)
**With Translation**
![Screenshot from 2020-11-02 18-51-39](https://user-images.githubusercontent.com/38663217/97873080-1d7eec80-1d3d-11eb-898b-fcd7b090bcaa.png)

vikram-raj commented on frontend/packages/container-security/src/components/summary.tsx:
```diff
@@ -77,19 +107,19 @@ export const SecurityBreakdownPopup: React.FC<SecurityBreakdownPopupProps> = ({
-                  to={`/k8s/all-namespaces/${referenceForModel(ImageManifestVulnModel)}`}
-                  aria-label="View all"
-                >
+                <Link to={baseVulnListUrl} aria-label="View all">
```
Add translation for `View All`.

nemesis09 replied:
done

vikram-raj commented on frontend/public/components/dashboard/project-dashboard/status-card.tsx:
```diff
@@ -1,22 +1,57 @@
+      ),
+    [subsystemExtensions],
+  ) as DashboardsOverviewHealthResourceSubsystem;
+  const namespace = useSelector((state: RootState) => getActiveNamespace(state));
```
Isn't we can use `obj.metadata.name` instead of `useSelector` for the namespace?

nemesis09 replied:
yes, `namespace` is available in `obj.name`
updated to use `obj.name` for namespace
Thanks.

vikram-raj commented on frontend/packages/container-security/src/components/summary.tsx:
```diff
@@ -44,26 +49,55 @@ export const quayURLFor = (vuln: ImageManifestVuln) => {
-        registries are not scanned.
+        {namespace
+          ? t(
+              `container-security~This Project's container images from Quay are analyzed to identify vulnerabilities. Images from other registries are not scanned.`,
```
```suggestion
              `container-security~This project's container images from Quay are analyzed to identify vulnerabilities. Images from other registries are not scanned.`,
```

nemesis09 replied:
done

vikram-raj reviewed (commented):
/lgtm

nemesis09 commented:
cc @openshift/team-devconsole-ux @itsptk

invincibleJai commented on frontend/packages/container-security/src/components/__tests__/summary.spec.tsx:
```diff
@@ -0,0 +1,134 @@
+    highVuln,
+    _.set(_.cloneDeep(highVuln), 'metadata.namespace', 'openshift-marketplace'),
+  ];
+  const i18nModule = 'container-security~';
```
nit: 
```suggestion
  const i18nNS = 'container-security~';
```

nemesis09 replied:
done

invincibleJai commented on frontend/packages/container-security/src/components/summary.tsx:
```diff
@@ -44,26 +49,55 @@ export const quayURLFor = (vuln: ImageManifestVuln) => {
+    } = v.status;
+    const totalCount = highCount + mediumCount + lowCount + unknownCount;
+    return n
+      ? t('container-security~{{fixableCount}} of {{totalCount}} fixable', {
```
can make use of number formatter here and wherever we use number
```suggestion
      ? t('container-security~{{fixableCount, number}} of {{totalCount, number}} fixable', {
```

nemesis09 replied:
done

invincibleJai commented on frontend/packages/container-security/src/plugin.ts:
```diff
@@ -96,7 +97,7 @@ const plugin: Plugin<ConsumedExtensions> = [
     type: 'Dashboards/Overview/Health/Resource',
     properties: {
-      title: 'Image Vulnerabilities',
+      title: i18n.t('container-security~Image Vulnerabilities'),
```
avoid use of global use of i18n, can leave `plugin.ts` for now

nemesis09 replied:
done

invincibleJai commented on frontend/public/components/dashboard/dashboards-page/cluster-dashboard/health-item.tsx:
```diff
@@ -226,10 +227,14 @@ export const PrometheusHealthItem = withDashboardResources<PrometheusHealthItemP
+    'imageManifestVuln.namespace',
+    namespace,
+  );
+  const resourcesResult = useK8sWatchResources(resourcesWithNamespace);
```
nit:can we add type here?

nemesis09 replied:
added type

invincibleJai commented on frontend/public/components/dashboard/project-dashboard/status-card.tsx:
```diff
@@ -1,22 +1,56 @@
 import * as React from 'react';
+// FIXME upgrading redux types is causing many errors at this time
+// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+// @ts-ignore
```
Why we need this?

nemesis09 replied:
this is not needed anymore. removed. Thanks.

invincibleJai commented on frontend/packages/container-security/src/components/summary.tsx:
```diff
@@ -19,10 +21,13 @@ export const securityHealthHandler: ResourceHealthHandler<WatchImageVuln> = ({
 
   if (loadError) {
-    return { state: HealthState.UNKNOWN, message: 'Not available' };
+    return { state: HealthState.UNKNOWN, message: i18n.t('container-security~Not available') };
```
avoid global use of `i18n`

nemesis09 replied:
removed global `i18n`

invincibleJai commented on frontend/public/components/dashboard/dashboards-page/cluster-dashboard/health-item.tsx:
```diff
@@ -226,10 +227,14 @@ export const PrometheusHealthItem = withDashboardResources<PrometheusHealthItemP
   const { title, resources, healthHandler, popupComponent, popupTitle } = subsystem;
-  const resourcesResult = useK8sWatchResources(resources);
-
+  const resourcesWithNamespace = _.set(
+    _.cloneDeep(resources),
+    'imageManifestVuln.namespace',
+    namespace,
+  );
```
can we have it as below?

```suggestion
const { imageManifestVuln : { kind : resourceKind} } = resources;
const resourcesWithNamespace = React.useMemo(() => {
  return {
        imageManifestVuln: {
            isList: true
            kind: resourceKind
            namespaced : true,
            namespace,
           }
      }
}, [resourceKind , namespace]);
```

nemesis09 replied:
updated it to
```
const { kind : resourceKind} = resources?.imageManifestVuln ?? {};
const resourcesWithNamespace = React.useMemo(() => {
  return {
        imageManifestVuln: {
            isList: true
            kind: resourceKind
            namespaced : true,
            namespace,
           }
      }
}, [resourceKind , namespace]);
```
to account for undefined and null values

nemesis09 replied:
updated it to 
`const resourcesWithNamespace: WatchK8sResources<ResourcesObject> = React.useMemo(() => {
    return {
      ...resources,
      ...(resources.imageManifestVuln && {
        imageManifestVuln: { ...resources.imageManifestVuln, namespace },
      }),
    };
  }, [resources, namespace]);`
as the previous implementation excluded the attributes other than `imageManifestVuln` and may cause errors in other components using this.

itsptk commented:
> ![Screenshot from 2020-11-03 20-03-26](https://user-images.githubusercontent.com/38663217/97998126-bbdb8280-1e0f-11eb-8566-3b5f7ef10195.png)

@nemesis09 It looked like this popover code might be shared with the cluster dashboard version, which is great. Can we also use the '52 of 52 fixable' format in the Vulnerabilities column on the cluster dashboard as well, since that is a little more useful and then they are consistent? 

Also ideally the project version of this popover would have text that starts with "This project's container images from quay..." like the ux design but not super critical.

vikram-raj commented:
/lgtm cancel 

My [PR](https://github.com/openshift/console/pull/6854) got merged. Please update the link of the vuln. It should redirect to the Vulnerabilities tab instead of the default list page of IMV.

nemesis09 commented:
@itsptk 
updated the text for admin cluster overview
I've also updated the links to redirect to Vulnerabilities tab in project details since the [PR](https://github.com/openshift/console/pull/6854)  has been merged.
Link changes ->
`Chart Donut` - 
**in project details**: redirects to the `Vulnerabilities` tab of the project details with the list of all vulnerabilities in the project/namespace
**in cluster overview(admin)**: redirects to the default details page with all vulnerabilites

`Impact` -
**in cluster overview(admin)**: redirects to the default details page with the selected vulnerability

`Image` -
**in project details**: redirects to the `Vulnerabilities` tab for the project with the selected vulnerability

`View All`- 
i**n project details**: redirects to the `Vulnerabilities` tab of the project details with the list of all vulnerabilities in the project/namespace
**in cluster overview(admin)**: redirects to the default details page with all vulnerabilites

I've attached a gif in the PR description for reference.
please take a look.

nemesis09 commented:
> /lgtm cancel
> 
> My [PR](https://github.com/openshift/console/pull/6854) got merged. Please update the link of the vuln. It should redirect to the Vulnerabilities tab instead of the default list page of IMV.

updated links

vikram-raj commented on frontend/packages/container-security/src/components/__tests__/summary.spec.tsx:
```diff
@@ -0,0 +1,132 @@
+  };
+});
+
+const highVuln: ImageManifestVuln = {
```
Can we use `fakeVulnFor` data define in `bad-pods.ts`?

nemesis09 replied:
done

vikram-raj commented on frontend/public/components/dashboard/project-dashboard/status-card.tsx:
```diff
@@ -5,18 +5,49 @@ import DashboardCardHeader from '@console/shared/src/components/dashboard/dashbo
+  const subsystemExtensions = useExtensions<DashboardsOverviewHealthSubsystem>(
+    isDashboardsOverviewHealthSubsystem,
+  );
+  const subsystem = React.useMemo(
+    () =>
+      filterSubsystems(subsystemExtensions).find(
+        (s) => s.properties.title === 'Image Vulnerabilities',
+      ),
+    [subsystemExtensions],
+  ) as DashboardsOverviewHealthResourceSubsystem;
```
WDYT?
```suggestion
  const subsystem: DashboardsOverviewHealthResourceSubsystem = React.useMemo(
    () =>
      filterSubsystems(subsystemExtensions).find(
        (s) => s.properties.title === 'Image Vulnerabilities',
      ),
    [subsystemExtensions],
  );
```

nemesis09 replied:
done

vikram-raj reviewed (commented):
Verified.
/lgtm

itsptk commented:
Thanks @nemesis09 

> `Impact` -
> **in cluster overview(admin)**: redirects to the default details page with the selected vulnerability
> 
> `Image` -
> **in project details**: redirects to the `Vulnerabilities` tab for the project with the selected vulnerability

So to confirm, a '3 namespaces' link in the **Impact** column for cluster overview would take the user to the vulnerabilities page, but filtered by that image, so they would see the 3 IMVs that relate to that image? (one per namespace.)

For the **Image** case in project details, this link could actually take the user directly into the details of that IMV for the current project, since there will only ever be 1 IMV that relates (the one in the current project.) This would be a change from what I saw, in that it was taking the user to the vulnerabilities tab in the project, with that list filtered and showing the single IMV in the list. Can be seen in this marvel: https://marvelapp.com/prototype/b8gddd8/screen/73052101

nemesis09 commented:
> Thanks @nemesis09
> 
> > `Impact` -
> > **in cluster overview(admin)**: redirects to the default details page with the selected vulnerability
> > `Image` -
> > **in project details**: redirects to the `Vulnerabilities` tab for the project with the selected vulnerability
> 
> So to confirm, a '3 namespaces' link in the **Impact** column for cluster overview would take the user to the vulnerabilities page, but filtered by that image, so they would see the 3 IMVs that relate to that image? (one per namespace.)
> 
> For the **Image** case in project details, this link could actually take the user directly into the details of that IMV for the current project, since there will only ever be 1 IMV that relates (the one in the current project.) This would be a change from what I saw, in that it was taking the user to the vulnerabilities tab in the project, with that list filtered and showing the single IMV in the list. Can be seen in this marvel: https://marvelapp.com/prototype/b8gddd8/screen/73052101

updated the link for image in the popover
PTAL

itsptk commented:
> updated the link for image in the popover
> PTAL

Looks great, thanks!

invincibleJai commented:
Thanks @nemesis09 , Verified the changes looks good

invincibleJai commented:
/lgtm

christianvogt commented:
/approve

vikram-raj commented:
/retest
